### ISSUE TITLE
Consider logging assembly loads into the binlog

### ISSUE THREAD

**Issue by KirillOsenkov** [dotnet/msbuild#6303]
We often have issues when MSBuild binaries are loaded from the GAC etc. Or NuGet/SDK binaries loaded from the wrong location. I think it would be beneficial to subscribe to AppDomain.CurrentDomain.AssemblyLoad and log all assembly loads, including full path to the .dll, assembly name, MVID, and maybe some attributes such as AssemblyFileVersion (which often contains the Git SHA).

**Comment by KirillOsenkov** [comment]
We could also show a warning/error if any assembly from the interesting list is loaded from the GAC.

**Comment by KirillOsenkov** [comment]
Related: https://github.com/dotnet/msbuild/issues/584

**Comment by KirillOsenkov** [comment]
oof, good questions. Also for desktop there's also AppDomains (MarkupCompile tooling) that I suggest we pretend doesn't exist for the purposes of this bug ;)

I wish there was a "node initialized for build" point in time where we could enumerate all assemblies loaded into the node process at the beginning. But I can't think of one.

Barring that, we could enumerate all loaded assemblies in the central node at the beginning of the build, but it's of limited used I guess (since we can already infer which MSBuild we're using because we log the path to MSBuild.exe). So we would be getting pretty much the same predictable list every time.

More interesting is the assemblies loaded by tasks and analyzers, i.e. all third-party assemblies. So maybe for starters don't worry about assemblies that are already loaded, and just focus on logging the ones loaded dynamically "at runtime". Ideally if there was a way to associate the assembly load event with a task that triggered it (maybe there's a "CurrentBuildContext" somewhere, so we could log a new event with the Node Id, Project Id, Target Id and Task Id, so we could attribute the load to a particular task running).

Also consider if we're running in a non-command-line scenario (doubt we can run inside the devenv process, maybe for evaluation?) Should we be logging assembly loads during evaluation? So many questions.

**Comment by KirillOsenkov** [comment]
An interesting thought in terms of implementation:

we could subscribe to AssemblyLoad in the task runner before a task starts executing, and unsubscribe after it's finished, and log all loads we saw while the task was running. This way we know exactly what the BuildEventContext is for the task, so messages will be nicely hanging off the task.

We're probably going to need to bump the binlog format for this, and introduce a new AssemblyLoadEventArgs deriving from MessageEventArgs. Never a fan of bumping the binlog format, but appears justified in this case.

Are analyzers always ran as part of a task? Will we catch assemblies loaded by analyzers? What about custom tasks?

**Comment by baronfel** [comment]
I thought Analyzers were passed directly to the relevant compiler calls (Csc, etc), then the compiler itself was on the hook for loading those (and all of their dependencies). So as a result, no analyzers should be loaded as part of the build evaluation/execution itself, only as part of the compiler invocation.

**Comment by KirillOsenkov** [comment]
Right, I thought so. Typically we should be shelling out to the compiler server process. But I was wondering if there are cases where stuff loads in-proc. Maybe like dotnet format, or linters or something. Just brainstorming. For now it feels like assemblies can be loaded either for custom tasks or during a task execution. During experimentation we should probably subscribe to everything and log all loads, and run it on a few builds to see what the typical scenarios are.
